src/crimson/os/seastore/omap_manager/btree/btree_omap_manager.h

```diff
@@ -0,0 +1,63 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+#include <boost/intrusive_ptr.hpp>
+#include <boost/smart_ptr/intrusive_ref_counter.hpp>
+#include <seastar/core/future.hh>
+
+#include "include/ceph_assert.h"
+#include "crimson/osd/exceptions.h"
+
+#include "crimson/os/seastore/omap_manager.h"
+#include "crimson/os/seastore/omap_manager/btree/omap_btree_node.h"
+#include "crimson/os/seastore/seastore_types.h"
+#include "crimson/os/seastore/transaction_manager.h"
+
+namespace crimson::os::seastore::omap_manager {
+
+class BtreeOMapManager : public OMapManager {
+  TransactionManager &tm;
+
+  omap_context_t get_omap_context(omap_root_t &omap_root, Transaction &t) {
+    return omap_context_t{omap_root, tm, t};
+  }
+
+  /* get_omap_root
+   *
+   * load omap tree root node
+   */
+  using get_root_ertr = TransactionManager::read_extent_ertr;
+  using get_root_ret = get_root_ertr::future<OMapNodeRef>;
+  get_root_ret get_omap_root(omap_root_t &omap_root, Transaction &t);
+
+  using insert_key_ertr = TransactionManager::read_extent_ertr;
+  using insert_key_ret = insert_key_ertr::future<std::pair<std::string, std::string>>;
+  insert_key_ret insert_key(omap_root_t &omap_root, Transaction &t,
+                            OMapNodeRef extent, std::string &key,
+                            std::string &val);
+
+public:
+  explicit BtreeOMapManager(TransactionManager &tm);
+
+  initialize_omap_ret initialize_omap(Transaction &t) final;
+
```

Comment:
More generally, I'm worried about taking all of these values by reference.  Your only current user uses unsafe_get0() to ensure that the operation completes before the stack-allocated strings lifetimes expire, but in general users of this interface will need to go to some effort to ensure that.  You should at a minimum add a comment at the top of this file warning users about that.

Reply:
I am also considering use value copy or reference,  value copy will copy many times in each key and value transfer.   I put a  reminder in omap_manager.h.  if you think value copy will be better, I can change it.

Reply:
That's fine, but add comments explaining that the caller *must* ensure that the referenced value *lives until the future resolves*.

Also, actually, Transactions have to be passed by reference, so you can ignore that.  I'm worried really about strings, sets, etc.